code/obj/machinery/pipe/pipe_dispenser.dm

```diff
@@ -11,62 +21,87 @@ TYPEINFO(/obj/machinery/disposal_pipedispenser)
 	anchored = 1
 	deconstruct_flags = DECON_SCREWDRIVER | DECON_WRENCH | DECON_CROWBAR | DECON_WELDER | DECON_WIRECUTTERS
 
-/obj/machinery/disposal_pipedispenser/attack_hand(mob/user)
-	if(..())
-		return
-
-	var/dat = {"<b>Disposal Pipes</b><br><br>
-<A href='?src=\ref[src];dmake=0'>Pipe</A><BR>
-<A href='?src=\ref[src];dmake=1'>Bent Pipe</A><BR>
-<A href='?src=\ref[src];dmake=2'>Junction</A><BR>
-<A href='?src=\ref[src];dmake=3'>Y-Junction</A><BR>
-<A href='?src=\ref[src];dmake=4'>Trunk</A><BR>
-"}
-
-	user.Browse("<HEAD><TITLE>Disposal Pipe Dispenser</TITLE></HEAD><TT>[dat]</TT>", "window=pipedispenser")
-	return
-
-// 0=straight, 1=bent, 2=junction-j1, 3=junction-j2, 4=junction-y, 5=trunk
-
-
-/obj/machinery/disposal_pipedispenser/Topic(href, href_list)
-	if(..())
+	var/dispenser_ready = TRUE
+	var/dispenser_delay = 5 DECI SECONDS
+	var/mobile = FALSE
+	var/datum/action/bar/icon/timer
+
+/obj/machinery/disposal_pipedispenser/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "PipeDispenser")
+		ui.open()
+
+/obj/machinery/disposal_pipedispenser/ui_data(mob/user)
+	. = list(
+		"dispenser_ready" = src.dispenser_ready,
+		"windowName" = src.name,
+	)
+
+/obj/machinery/disposal_pipedispenser/ui_static_data(mob/user)
+	. = list(
+		"mobile" = src.mobile,
+		"max_disposal_pipes" = MAX_BUILD_DISPOSAL,
+	)
+	for (var/disposaltype in availdisposalpipes)
+		.["disposalpipes"] += list(list(
+			"disposaltype" = disposaltype,
+			"image" = getBase64ImgDisposal(disposaltype)
+			))
+
+/obj/machinery/disposal_pipedispenser/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	if (.)
 		return
-	src.add_dialog(usr)
-	src.add_fingerprint(usr)
-	if(href_list["dmake"])
-		var/p_type = text2num_safe(href_list["dmake"])
+	switch(action)
+		if("dmake")
+			if (!dispenser_ready)
+				boutput(ui.user, "<span class='alert'>The [src] isn't ready yet!</span>")
+				return
+			var/p_type = text2num_safe(availdisposalpipes[params["disposal_type"]])
+			if (isnull(p_type))
+				return
```

Comment:
Consider some sort of logging thing here. In theory the user shouldn't get here, but rather than just aborting silently this should be logged so that the bad state is easier to track down.

Reply:
Using stack_trace for this.